<a href="https://colab.research.google.com/github/Svitlana-Yarish/Yarish_Python-Assignment/blob/main/4_1_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 4.1

> Replace all TODOs with your code.
>
> Do not change any other code and do not add/remove cells!

## Sets

### Task 1

Create an empty set called 'my_set'.

In [1]:
my_set = set()

print(my_set)  # should print set()

set()


### Task 2

Let's turn a given sentence into a set of words.


In [3]:
sentence = "abc cba acd abc"

words = sentence.split()
words = set(words)

print(words)  # should be {'acd', 'cba', 'abc'} or in other order

{'acd', 'cba', 'abc'}


### Task 3

Now, let's remove from `words` all the words that are present in `sentence2`.

In [5]:
sentence2 = 'cba baa acc'
sentence2 = sentence2.split()
sentence2 = set(sentence2)
words = words - sentence2


print(words)  # should print {'acd', 'abc'} or {'abc', 'acd'}

{'acd', 'abc'}


### Task 4

Let's now play around with statistics and probability!

Generate `N` random numbers in the range `[1, N]` and keep track of a percent of unique numbers in the result.
Play around with different values of `N`; how does it affect the percentage of unique numbers? How do you think why?

In [6]:
from random import randint

N = 10000000

def unique_percentage(N: int) -> float:

    random_numbers = [randint(1, N) for _ in range(N)]  # генеруємо числа (через функцію randint)

    unique_numbers = set(random_numbers)  #  видалили дублікати

    return (len(unique_numbers) / N) * 100 # тут обчислюємо відсоткове значення за формулою

percentage = unique_percentage(N)

print(f"N = {N}: Відсоток унікальних чисел = {percentage:.2f}%")

N = 10000000: Відсоток унікальних чисел = 63.21%


For math geeks: if you set rather big `N`, you see that the percentage is slowly approaching ~63.21. It is a modification of a [Birthday problem](https://en.wikipedia.org/wiki/Birthday_problem), but a simpler and more relevant explanation is [here](https://math.stackexchange.com/questions/5775/how-many-bins-do-random-numbers-fill).


## Dictionaries

### Task 5

You are given a matrix `m`. Return a distinct list containing all the values that are present in at least `n` rows.

Hint: there can be duplicates in rows; perhaps we would want to eliminate them - how?

In [12]:
m = [
    [1, 1, 2, 3],
    [1, 2, 6, 7],
    [1, 7, 9, 9]
]

n = 2
res = []

value_count = {}

for row in m:
  unique_values = set(row)

  for value in unique_values:
    if value in value_count:
       value_count[value] += 1
    else:
      value_count[value] = 1

for value, count in value_count.items():
    if count >= n:
        res.append(value)

res.sort()

print(res)  # should print [1, 2, 7]

[1, 2, 7]
